66.md

```diff
@@ -0,0 +1,250 @@
+# NIP-66: Relay Discovery and Liveness Monitoring
+
+`draft` `optional`
+
+You want to find relays. You may want to discover relays based on criteria that's up to date. You may even want to ensure that you have a complete dataset. You probably want to filter relays based on their reported liveness.
+
+In its purest form:
+
+```json
+{
+  "kind": 30166,
+  "created_at": 1722173222,  
+  "content": "{}",
+  "tags": [
+    [ "d", "wss://somerelay.abc/" ]
+  ],
+  "pubkey": "<pubkey>", 
+  "sig": "<signature>",
+  "id": "<eventid>"
+}
+```
+
+This event signals that the relay at `wss://somerelay.abc/` was reported "online" by `<pubkey>` at timestamp `1722173222`. This event **MAY** be extended upon to include more information.
+
+## Kinds
+`NIP-66` defines two (2) event kinds, `30166` and `10166`
+
+| kind  | name                       | description                                                                             |
+|-------|----------------------------|-----------------------------------------------------------------------------------------|
+| [30166](#k30166) | Relay Discovery      | A PRE that is published by a monitor when a relay is online |
+| [10166](#k10166) | Relay Monitor Announcement | An RE that stores data that signals the intent of a pubkey to monitor relays and publish `30166` events at a regular _frequency_ |
+
+## Ontology
+- `Relay Operator`: someone who operates a relay 
+- `Monitor`: A pubkey that monitors relays and publishes `30166` events at the frequency specified in their `10166` event.
+- `Ad-hoc Monitor`: A pubkey that monitors relays and publishes `30166` events at an irregular frequency.
+- `Monitor Service`: A group or individual that monitors relays using one or more `Monitors`.
+- `Check`: a specific data point that is tested or aggregated by a monitor.
+
+## `30166`: "Relay Discovery" <a id="k30166"></a>
+
+### Summary
+`30166` is a `NIP-33` Parameterized-Replaceable Event [PRE], referred to as a "Relay Discovery" event. These events are optimized with a small footprint for protocol-level relay Discovery.
+
+### Purpose 
+Discovery of relays over nostr. 
+
+### Schema 
+
+#### Content 
+`30166` content fields **MAY** include the stringified JSON of the relay's NIP-11 informational document. This data **MAY** be provided for informational purposes only. 
+
+#### `created_at`
+The `created_at` field in a NIP-66 event should reflect the time when the relay liveness (and potentially other data points) was checked.
+
+#### `tags`
+
+##### Meta Tags (unindexed)
+- `rtt-open` The relay's open **round-trip time** in milliseconds.
+- `rtt-read` The relay's read **round-trip time** in milliseconds.
+- `rtt-write` The relay's write **round-trip time** in milliseconds.
```

Comment:
How are these measured?

Reply:
read: `time.start` subscribe, subscription fulfilled or EOSE recieved (whichever first) `time.end`
write: `time.start` send event, ok received `time.end` 

Notes:
1. Both of these values are more useful when represented as an SMA; some relay implementations are slower than others. 
2. Both of these values can confirm relay behaviors when cross-referencing other relay attributes. 
3. These values cannot be universally applied since relays treat all users differently. 
4. Write tests are particularly questionable. I'm still using the testing methodology that I inherited from previous relay checking attempts.

Reply:
After some thought, `read` and `write` should probably be removed from this NIP. For these kinds of tests to be meaningful in the present-day nostr climate requires more context. 

I wrote an [auditor](https://github.com/sandwichfarm/auditor), similar to to mike's relaytester, and it seems clear that read checks will be useful, but with very specific parameters to highlight strengths and weaknesses between relays. 

How to propagate or aggregate this data is still unclear, but it is very clear (and has been for a while tbh) that the simple read and write checks of yester-year provide little to no value anymore. I even completely disregard them in the new nostr.watch NIP-66 client.